File: Advanced Robotics Course/Chapter6/Lesson1/Chapter6_Lesson1.cpp
#include <Eigen/Dense>
#include <random>
#include <iostream>

using Vector3 = Eigen::Vector3d;
using Matrix3 = Eigen::Matrix3d;

// Time step
const double dt = 0.1;

Vector3 f(const Vector3& x, const Eigen::Vector2d& u)
{
    double px = x(0);
    double py = x(1);
    double phi = x(2);
    double v = u(0);
    double omega = u(1);

    Vector3 xn;
    xn(0) = px + v * std::cos(phi) * dt;
    xn(1) = py + v * std::sin(phi) * dt;
    xn(2) = phi + omega * dt;
    return xn;
}

int main()
{
    // Process noise covariance
    Matrix3 Q = Matrix3::Zero();
    Q(0,0) = 0.01 * 0.01;
    Q(1,1) = 0.01 * 0.01;
    Q(2,2) = std::pow(M_PI / 180.0, 2); // 1 deg in rad

    std::mt19937 gen(0);
    std::normal_distribution<double> standard_normal(0.0, 1.0);

    auto sample_process_noise = [&]() {
        Vector3 z;
        for (int i = 0; i < 3; ++i)
            z(i) = standard_normal(gen);
        // Cholesky factor of Q
        Eigen::LLT<Matrix3> llt(Q);
        Matrix3 L = llt.matrixL();
        return L * z;
    };

    Vector3 x;
    x << 0.0, 0.0, 0.0;
    Eigen::Vector2d u;
    u << 1.0, 0.2;

    const int T = 20;
    for (int t = 0; t < T; ++t) {
        Vector3 x_nom = f(x, u);
        Vector3 w = sample_process_noise();
        x = x_nom + w;
    }

    std::cout << "Final stochastic state: " << x.transpose() << std::endl;

    // In a ROS2 or ROS1 stack, such propagation would often be part of a node
    // using Eigen, tf2, and sensor_msgs for message-passing.
    return 0;
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter6/Lesson1/Chapter6_Lesson1.java
import java.util.Random;

public class StochasticUnicycle {

    static final double DT = 0.1;
    static final Random rng = new Random(42L);

    // x = [x, y, phi]
    static double[] f(double[] x, double[] u) {
        double px = x[0];
        double py = x[1];
        double phi = x[2];
        double v = u[0];
        double omega = u[1];

        return new double[] {
            px + v * Math.cos(phi) * DT,
            py + v * Math.sin(phi) * DT,
            phi + omega * DT
        };
    }

    static double[] sampleProcessNoise(double sigmaX, double sigmaY, double sigmaPhi) {
        return new double[] {
            sigmaX * rng.nextGaussian(),
            sigmaY * rng.nextGaussian(),
            sigmaPhi * rng.nextGaussian()
        };
    }

    public static void main(String[] args) {
        double[] x = new double[] {0.0, 0.0, 0.0};
        double[] u = new double[] {1.0, 0.2};

        // Standard deviations of process noise
        double sigmaX = 0.01;
        double sigmaY = 0.01;
        double sigmaPhi = Math.toRadians(1.0);

        for (int t = 0; t < 20; ++t) {
            double[] xNom = f(x, u);
            double[] w = sampleProcessNoise(sigmaX, sigmaY, sigmaPhi);
            for (int i = 0; i < 3; ++i) {
                x[i] = xNom[i] + w[i];
            }
        }

        System.out.printf("Final stochastic state: x=%.3f, y=%.3f, phi=%.3f%n",
                          x[0], x[1], x[2]);
    }
}
      
